armory/metrics/task.py

```diff
@@ -640,6 +869,88 @@ def video_tracking_mean_success_rate(y, y_pred):
+    iou_threshold=0.5,
+    class_list=None,
+    mean=True,
+    increment=0.1,
+):
+    """
+    Mean average precision for adversarial object detection, organizing results according to
+    how far boxes are from the patch.  The motivation is that a patch is most effective against
+    nearby objects, and reporting mAP over a range of distances from the patch will give additional
+    insight into the attack's behavior.
+
+    This function uses GIoU as a proxy for distance to patch.  GIoU ranges from -1 to 1, with positive
+    values indicating overlap and smaller negative values indicating a greater distance.
+
+    This function returns a dictionary, where the keys are the GIoU thresholds, and the values are mAP
+    for all objects within that range of the patch.
+
+    y_list (list): of length equal to the number of input examples. Each element in the list
+        should be a dict with "labels" and "boxes" keys mapping to a numpy array of
+        shape (N,) and (N, 4) respectively where N = number of boxes.
+    y_pred_list (list): of length equal to the number of input examples. Each element in the
+        list should be a dict with "labels", "boxes", and "scores" keys mapping to a numpy
+        array of shape (N,), (N, 4), and (N,) respectively where N = number of boxes.
+    y_patch_metadata_list (list): of length equal to the number of input examples.  Each element
+        is a dict with a "gs_coords" key mapping to a numpy array of shape (4,2).
+    class_list (list, optional): a list of classes, such that all predictions and ground-truths
+        with labels NOT in class_list are to be ignored.
+    mean: if False, returns a dict mapping each class to its average precision (AP)
+        if True, calls `mean_ap` on the AP dict and returns a encapsulating dict:
+        {'class': {<class_0>: <class_0_AP>, ...}, 'mean': <mean_AP>}
+    increment: the amount to increment the distance threshold for each reported mAP value
+    """
+
+    y_distances_list = []
+    y_pred_distances_list = []
+    result = {}
+
+    for y, y_pred, metadata in zip(y_list, y_pred_list, y_patch_metadata_list):
+        # For each image, use GIoU as proxy for distance between boxes and patch.
+        # GIoU is positive if there is overlap.
+        patch = metadata["gs_coords"]
+        y_distances = [
+            _generalized_intersection_over_union(box, patch) for box in y["boxes"]
+        ]
+        pred_distances = [
+            _generalized_intersection_over_union(box, patch) for box in y_pred["boxes"]
+        ]
+        y_distances_list.append(y_distances)
+        y_pred_distances_list.append(pred_distances)
+
+    # Compute mAP for boxes restricted to a minimum GIoU
+    for threshold in np.arange(0, -1 - increment, -increment):
+        # Start with 0 -- all boxes that overlap the patch
+        y_list_ = [
+            {
+                "boxes": y["boxes"][y_d > threshold],
+                "labels": y["labels"][y_d > threshold],
```

Comment:
It would also be nice to produce metrics for `y_d < threshold`.

Reply:
Do you think output like this is sufficiently clear and useful?  Note the latter subdict does not include -0.9 because there were no boxes that distant.
```
adversarial_AP_per_class_by_giou_from_patch on benign examples w.r.t. ground truth labels: {
'cumulative_by_min_giou': {
        0.0: {'mean': 0.005, 'class': {1: 0.0, 2: 0.01}}, 
        -0.1: {'mean': 0.065, 'class': {1: 0.05, 2: 0.08}},
        -0.2: {'mean': 0.15000000000000002, 'class': {1: 0.07, 2: 0.23}}, 
        -0.30000000000000004: {'mean': 0.185, 'class': {1: 0.13, 2: 0.24}}, 
        -0.4: {'mean': 0.2, 'class': {1: 0.15, 2: 0.25}}, 
        -0.5: {'mean': 0.265, 'class': {1: 0.2, 2: 0.33}}, 
        -0.6000000000000001: {'mean': 0.285, 'class': {1: 0.22, 2: 0.35}}, 
        -0.7000000000000001: {'mean': 0.305, 'class': {1: 0.25, 2: 0.36}}, 
        -0.8: {'mean': 0.315, 'class': {1: 0.27, 2: 0.36}}, 
        -0.9: {'mean': 0.355, 'class': {1: 0.29, 2: 0.42}}, 
        -1.0: {'mean': 0.355, 'class': {1: 0.29, 2: 0.42}}
        }, 
'cumulative_by_max_giou': {
        0.0: {'mean': 0.505, 'class': {1: 0.4, 2: 0.61}}, 
        -0.1: {'mean': 0.63, 'class': {1: 0.58, 2: 0.68}}, 
        -0.2: {'mean': 0.645, 'class': {1: 0.6, 2: 0.69}}, 
        -0.30000000000000004: {'mean': 0.6799999999999999, 'class': {1: 0.65, 2: 0.71}}, 
        -0.4: {'mean': 0.655, 'class': {1: 0.6, 2: 0.71}}, 
        -0.5: {'mean': 0.6399999999999999, 'class': {1: 0.59, 2: 0.69}}, 
        -0.6000000000000001: {'mean': 0.675, 'class': {1: 0.67, 2: 0.68}}, 
        -0.7000000000000001: {'mean': 0.675, 'class': {1: 0.65, 2: 0.7}}, 
        -0.8: {'mean': 0.755, 'class': {1: 0.79, 2: 0.72}}
        }
}
```

Reply:
Yeah, I think this makes sense. I do wonder, however, if a histogram of GIoUs might be more intuitive to interpret. The cocoapi does this, for example, with APs computed at different groundtruth object mask areas (see "AP Across Scales" https://cocodataset.org/#detection-eval):
> In COCO, there are more small objects than large objects. Specifically: approximately 41% of objects are small (area < 322), 34% are medium (322 < area < 962), and 24% are large (area > 962). Area is measured as the number of pixels in the segmentation mask.